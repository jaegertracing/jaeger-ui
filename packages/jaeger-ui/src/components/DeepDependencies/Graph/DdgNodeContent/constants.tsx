// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { getUrl as getSearchUrl } from '../../../SearchTracePage/url';

// Constants for calculating circular node positioning
export const FONT_SIZE = 14;
export const FONT = `${FONT_SIZE}px Helvetica Nueue`;
export const LINE_HEIGHT = 1.5;
export const OP_PADDING_TOP = 5;
export const PROGRESS_BAR_STROKE_WIDTH = 15;
export const RADIUS = 75;
export const WORD_RX = /\W*\w+\W*/g;

// While browsers suport URLs of unlimited length, many server clients do not handle more than this max
// export const MAX_LENGTH = 2083;
// get unpreditable, unreproducable results past 7400 characters
export const MAX_LENGTH = 7000;
export const MAX_LINKED_TRACES = 35;
export const MIN_LENGTH = getSearchUrl().length;
export const PARAM_NAME_LENGTH = '&traceID='.length;
