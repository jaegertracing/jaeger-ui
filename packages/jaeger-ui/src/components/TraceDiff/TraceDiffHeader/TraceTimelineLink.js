// @flow

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { getUrl } from '../../TracePage/url';
import NewWindowIcon from '../../common/NewWindowIcon.js';

type PropsType = {
  traceID: string,
};

function stopPropagation(event: SyntheticMouseEvent<HTMLAnchorElement>) {
  event.stopPropagation();
}

export default function TraceTimelineLink({ traceID }: PropsType) {
  return (
    <a href={getUrl(traceID)} onClick={stopPropagation} rel="noopener noreferrer" target="_blank">
      <NewWindowIcon />
    </a>
  );
}
