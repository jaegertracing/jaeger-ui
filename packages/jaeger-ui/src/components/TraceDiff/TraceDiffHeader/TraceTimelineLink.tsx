// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { getUrl } from '../../TracePage/url';
import NewWindowIcon from '../../common/NewWindowIcon';
import { getTargetEmptyOrBlank } from '../../../utils/config/get-target';

type PropsType = {
  traceID: string;
};

function stopPropagation(event: React.MouseEvent<HTMLAnchorElement>) {
  event.stopPropagation();
}

export default function TraceTimelineLink({ traceID }: PropsType) {
  return (
    <a
      href={getUrl(traceID)}
      onClick={stopPropagation}
      rel="noopener noreferrer"
      target={getTargetEmptyOrBlank()}
    >
      <NewWindowIcon />
    </a>
  );
}
