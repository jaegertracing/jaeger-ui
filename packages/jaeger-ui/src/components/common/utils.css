/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.u-width-100 {
  width: 100%;
}

.u-flex-1 {
  flex: 1;
}

.u-mt-vast {
  margin-top: 13rem;
}

.u-cursor-pointer {
  cursor: pointer;
}

.u-tx-muted {
  color: var(--tx-color-muted);
}

.u-tx-inherit {
  color: inherit;
}

.u-tx-ellipsis {
  text-overflow: ellipsis;
}

.u-align-icon {
  margin: -0.2rem 0.25rem 0 0;
}

.u-simple-card {
  background-color: #f5f5f5;
  border: 1px solid #e6e6e6;
  padding: 1rem;
}

/* simple-scrollbars */
.u-simple-scrollbars::-webkit-scrollbar {
  -webkit-appearance: none;
  width: 8px;
  height: 8px;
}

.u-simple-scrollbars::-webkit-scrollbar-track {
  background: rgba(0, 0, 0, 0.03);
}

.u-simple-scrollbars::-webkit-scrollbar-track:vertical {
  border-left: 1px solid rgba(0, 0, 0, 0.13);
}

.u-simple-scrollbars::-webkit-scrollbar-track:horizontal {
  border-top: 1px solid rgba(0, 0, 0, 0.13);
}

.u-simple-scrollbars::-webkit-scrollbar-thumb {
  background: rgba(0, 0, 0, 0.2);
}

.u-simple-scrollbars:hover::-webkit-scrollbar-thumb {
  background: rgba(0, 0, 0, 0.35);
}

.u-simple-scrollbars::-webkit-scrollbar-thumb:window-inactive {
  background: rgba(0, 0, 0, 0.15);
}

/* Remove the padding around ant-design's popover, unfortunate but very handy */
.u-rm-popover-title-padding .ant-popover-title,
.u-rm-popover-content-padding .ant-popover-inner-content {
  padding: 0;
}

/* minimap for plexus */

.u-miniMap {
  align-items: flex-end;
  bottom: 1rem;
  display: flex;
  left: 1rem;
  position: absolute;
  z-index: 1;
}

.u-miniMap > .plexus-MiniMap--item {
  background: #999;
  box-shadow: 0px 0px 5px rgba(0, 0, 0, 0.3);
  margin-right: 1rem;
  position: relative;
}

/* Add the border as a pseudo element so the inner content doesn't mask it */
.u-miniMap > .plexus-MiniMap--item::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  outline: 1px solid #777;
}

.u-miniMap > .plexus-MiniMap--map {
  /* dynamic widht, height */
  box-sizing: content-box;
  cursor: not-allowed;
}

.u-miniMap .plexus-MiniMap--mapActive {
  /* dynamic: width, height, transform */
  background: #ccc;
  position: absolute;
}

.u-miniMap > .plexus-MiniMap--button {
  background: #ccc;
  color: #888;
  cursor: pointer;
  font-size: 1.6em;
  line-height: 0;
  padding: 0.1rem;
}

.u-miniMap > .plexus-MiniMap--button:hover {
  background: #ddd;
}
