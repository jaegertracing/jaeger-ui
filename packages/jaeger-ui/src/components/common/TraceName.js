// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import BreakableText from '../common/BreakableText';
import LoadingIndicator from '../common/LoadingIndicator';
import { fetchedState, FALLBACK_TRACE_NAME } from '../../constants';

import type { FetchedState } from '../../types';
import type { ApiError } from '../../types/api-error';

import './TraceName.css';

type Props = {
  className?: string,
  error?: ?ApiError,
  state?: ?FetchedState,
  traceName?: ?string,
};

export default function TraceName(props: Props) {
  const { className, error, state, traceName } = props;
  const isErred = state === fetchedState.ERROR;
  let title = traceName || FALLBACK_TRACE_NAME;
  let errorCssClass = '';
  if (isErred) {
    errorCssClass = 'is-error';
    let titleStr = '';
    if (error) {
      titleStr = typeof error === 'string' ? error : error.message || String(error);
    }
    if (!titleStr) {
      titleStr = 'Error: Unknown error';
    }
    title = <BreakableText text={titleStr} />;
  } else if (state === fetchedState.LOADING) {
    title = <LoadingIndicator small />;
  } else {
    const text = traceName || FALLBACK_TRACE_NAME;
    title = <BreakableText text={text} />;
  }
  return <span className={`TraceName ${errorCssClass} ${className || ''}`}>{title}</span>;
}

TraceName.defaultProps = {
  className: '',
  error: null,
  state: null,
  traceName: null,
};
