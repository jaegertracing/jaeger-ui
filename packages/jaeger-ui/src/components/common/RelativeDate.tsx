// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import dayjs from 'dayjs';

import { formatRelativeDate } from '../../utils/date';

type Props = {
  fullMonthName: boolean | undefined | null;
  includeTime: boolean | undefined | null;
  value: number | Date;
};

// TODO typescript doesn't understand text or null as react nodes
// https://github.com/Microsoft/TypeScript/issues/21699
export default function RelativeDate(props: Props): JSX.Element {
  const { value, includeTime, fullMonthName } = props;
  const m = dayjs.isDayjs(value) ? value : dayjs(value);
  const dateStr = formatRelativeDate(m, Boolean(fullMonthName));
  const timeStr = includeTime ? `, ${m.format('h:mm:ss a')}` : '';
  return <span>{`${dateStr}${timeStr}`}</span>;
}
