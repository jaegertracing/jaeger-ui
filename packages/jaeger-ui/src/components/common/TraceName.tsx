// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import BreakableText from './BreakableText';
import LoadingIndicator from './LoadingIndicator';
import { fetchedState, FALLBACK_TRACE_NAME } from '../../constants';

import { FetchedState, TNil } from '../../types';
import { ApiError } from '../../types/api-error';

import './TraceName.css';

type Props = {
  className?: string;
  error?: ApiError | TNil;
  state?: FetchedState | TNil;
  traceName?: string | TNil;
};

export default function TraceName(props: Props) {
  const { className, error, state, traceName } = props;
  const isErred = state === fetchedState.ERROR;
  let title: string | React.ReactNode = traceName || FALLBACK_TRACE_NAME;
  let errorCssClass = '';
  if (isErred) {
    errorCssClass = 'is-error';
    let titleStr = '';
    if (error) {
      titleStr = typeof error === 'string' ? error : error.message || String(error);
    }
    if (!titleStr) {
      titleStr = 'Error: Unknown error';
    }
    title = titleStr;
    title = <BreakableText text={titleStr} />;
  } else if (state === fetchedState.LOADING) {
    title = <LoadingIndicator small />;
  } else {
    const text = String(traceName || FALLBACK_TRACE_NAME);
    title = <BreakableText text={text} />;
  }
  return <span className={`TraceName ${errorCssClass} ${className || ''}`}>{title}</span>;
}
