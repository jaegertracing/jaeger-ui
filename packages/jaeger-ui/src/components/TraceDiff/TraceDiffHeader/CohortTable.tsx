// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Table, Tag } from 'antd';

import TraceTimelineLink from './TraceTimelineLink';
import RelativeDate from '../../common/RelativeDate';
import TraceName from '../../common/TraceName';
import { fetchedState } from '../../../constants';
import { formatDuration } from '../../../utils/date';

import { FetchedTrace, TNil } from '../../../types';

import './CohortTable.css';

type Props = {
  selection: Record<string, { label: string }>;
  current: string | TNil;
  cohort: FetchedTrace[];
  selectTrace: (traceID: string) => void;
};

const { Column } = Table;

const defaultRowSelection = {
  hideDefaultSelections: true,
  type: 'radio' as 'radio',
};

export const NEED_MORE_TRACES_MESSAGE = (
  <h3 key="msg" className="CohortTable--needMoreMsg">
    Enter a Trace ID or perform a search and select from the results.
  </h3>
);

export default class CohortTable extends React.PureComponent<Props> {
  getCheckboxProps = (record: FetchedTrace) => {
    const { current, selection } = this.props;
    const { id, state } = record;
    if (state === fetchedState.ERROR || (id in selection && id !== current)) {
      return { disabled: true };
    }
    return {};
  };

  render() {
    const { cohort, current, selection, selectTrace } = this.props;
    const rowSelection = {
      ...defaultRowSelection,
      getCheckboxProps: this.getCheckboxProps,
      // TODO: Antd Table believes onChange can be called with a string or number, but that seems wrong
      onChange: (ids: number[] | string[]) => selectTrace(ids[0] as string),
      selectedRowKeys: current ? [current] : [],
    };

    return [
      <Table
        key="table"
        size="middle"
        dataSource={cohort}
        rowKey="id"
        pagination={false}
        rowSelection={rowSelection}
      >
        <Column
          key="traceID"
          title=""
          dataIndex="id"
          render={value => <span className="u-tx-muted">{value && value.slice(0, 7)}</span>}
        />
        <Column
          key="traceName"
          title="Service &amp; Operation"
          sortOrder="descend"
          dataIndex="data.traceName"
          render={(_, record: FetchedTrace) => {
            const { data, error, id, state } = record;
            const { traceName = undefined } = data || {};
            const { label = undefined } = selection[id] || {};
            return (
              <React.Fragment>
                {label != null && (
                  <Tag key="lbl" className="ub-bold" color="#139999">
                    {label}
                  </Tag>
                )}
                <TraceName
                  key="name"
                  className="CohortTable--traceName"
                  error={error}
                  state={state}
                  traceName={traceName}
                />
              </React.Fragment>
            );
          }}
        />
        <Column
          title="Date"
          dataIndex="data.startTime"
          key="startTime"
          render={(value, record: FetchedTrace) =>
            record.state === fetchedState.DONE && (
              <RelativeDate fullMonthName includeTime value={value / 1000} />
            )
          }
        />
        <Column
          title="Duration"
          dataIndex="data.duration"
          key="duration"
          render={(value, record: FetchedTrace) =>
            record.state === fetchedState.DONE && formatDuration(value)
          }
        />
        <Column title="Spans" dataIndex="data.spans.length" key="spans" />
        <Column
          className="ub-tx-center"
          dataIndex="data.traceID"
          key="link"
          render={value => <TraceTimelineLink traceID={value} />}
        />
      </Table>,
      cohort.length < 2 && NEED_MORE_TRACES_MESSAGE,
    ];
  }
}
