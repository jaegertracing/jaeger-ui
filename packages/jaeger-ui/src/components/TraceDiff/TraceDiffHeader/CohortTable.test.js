// Copyright (c) 2019 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Table, Tag } from 'antd';

import CohortTable, { NEED_MORE_TRACES_MESSAGE } from './CohortTable';
import TraceTimelineLink from './TraceTimelineLink';
import RelativeDate from '../../common/RelativeDate';
import TraceName from '../../common/TraceName';
import { fetchedState } from '../../../constants';
import * as dateUtils from '../../../utils/date';

const { Column } = Table;

describe('CohortTable', () => {
  const cohort = [
    {
      data: {
        traceName: 'trace name 0',
      },
      error: 'api error',
      id: 'trace-id-0',
      state: fetchedState.ERROR,
    },
    {
      id: 'trace-id-1',
    },
    {
      id: 'trace-id-2',
    },
  ];
  const selectTrace = jest.fn();
  const props = {
    cohort,
    current: cohort[0].id,
    selection: {
      [cohort[0].id]: {
        label: 'selected index 0',
      },
    },
    selectTrace,
  };

  let formatDurationSpy;
  let wrapper;

  /**
   * Creates a new wrapper with default props and specified props. It is necessary to create a new wrapper
   * when props change because enzyme does not support wrapper.setProps for classes that render an array of
   * elements.
   *
   * @param {Object} [specifiedProps={}] - Props to set that are different from props defined above.
   * @returns {Object} - New wrapper.
   */
  function updateWrapper(specifiedProps = {}) {
    wrapper = shallow(<CohortTable {...props} {...specifiedProps} />);
  }

  function getRowRenderer(dataIndex) {
    return wrapper.find(Column).find(`[data-testid="${dataIndex}"]`).prop('render');
  }

  beforeAll(() => {
    formatDurationSpy = jest.spyOn(dateUtils, 'formatDuration');
  });

  beforeEach(() => {
    selectTrace.mockReset();
    formatDurationSpy.mockReset();
    updateWrapper();
  });

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('row selection', () => {
    let rowSelection;

    function updateRowSelection() {
      rowSelection = wrapper.find(Table).prop('rowSelection');
    }

    beforeEach(() => {
      updateRowSelection();
    });

    it('defaults selectedRowKeys to empty array', () => {
      updateWrapper({ current: undefined });
      updateRowSelection();
      expect(rowSelection.selectedRowKeys).toEqual([]);
    });

    it('calls props.selectTrace on row selection', () => {
      rowSelection.onChange([cohort[1].id, cohort[2].id], [cohort[1], cohort[2]]);
      expect(selectTrace).toHaveBeenCalledWith(cohort[1].id);
    });

    it('calculates checkbox props for selected and current record with error', () => {
      expect(rowSelection.getCheckboxProps(cohort[0])).toEqual({ disabled: true });
    });

    it('calculates checkbox props for selected and current record without error', () => {
      expect(rowSelection.getCheckboxProps({ ...cohort[0], state: fetchedState.DONE })).toEqual({});
    });

    it('calculates checkbox props for selected but not current record without error', () => {
      updateWrapper({
        selection: {
          ...props.selecetion,
          [cohort[1].id]: {
            label: 'selected index 1',
          },
        },
      });
      updateRowSelection();
      expect(rowSelection.getCheckboxProps(cohort[1])).toEqual({ disabled: true });
    });

    it('calculates checkbox props for not selected record', () => {
      expect(rowSelection.getCheckboxProps(cohort[1])).toEqual({});
    });
  });

  it('renders shortened id', () => {
    const idRenderer = getRowRenderer('id');
    const traceID = 'trace-id-longer-than-eight-characters';
    const renderedId = shallow(idRenderer(traceID));
    expect(renderedId.text()).toBe(traceID.slice(0, 7));
  });

  it('renders TraceName fragment when given complete data', () => {
    const traceNameColumnRenderer = getRowRenderer('traceName');
    const testTrace = cohort[0];
    const {
      id,
      error,
      state,
      data: { traceName },
    } = testTrace;
    const renderedTraceNameColumn = shallow(
      // traceNameRenderer returns a React Fragment, wrapper div helps enzyme
      <div>{traceNameColumnRenderer('unused argument', testTrace)}</div>
    );

    const tag = renderedTraceNameColumn.find(Tag);
    expect(tag.length).toBe(1);
    expect(tag.html().includes(props.selection[id].label)).toBe(true);

    const renderedTraceName = renderedTraceNameColumn.find(TraceName);
    expect(renderedTraceName.length).toBe(1);
    expect(renderedTraceName.props()).toEqual(
      expect.objectContaining({
        error,
        state,
        traceName,
      })
    );
  });

  it('renders TraceName fragment when given minimal data', () => {
    const traceNameColumnRenderer = getRowRenderer('traceName');
    const testTrace = cohort[1];
    const renderedTraceNameColumn = shallow(
      // traceNameRenderer returns a React Fragment, wrapper div helps enzyme
      <div>{traceNameColumnRenderer('unused argument', testTrace)}</div>
    );

    expect(renderedTraceNameColumn.find(Tag).length).toBe(0);
    expect(renderedTraceNameColumn.find(TraceName).length).toBe(1);
  });

  it('renders date iff record state is fetchedState.DONE', () => {
    const dateRenderer = getRowRenderer('startTime');
    const date = 1548689901403;

    expect(dateRenderer(date, { state: fetchedState.ERROR })).toBe(false);
    const renderedDate = dateRenderer(date, { state: fetchedState.DONE });
    expect(renderedDate.type).toBe(RelativeDate);
    expect(renderedDate.props).toEqual({
      fullMonthName: true,
      includeTime: true,
      value: date / 1000,
    });
  });

  it('renders duration iff record state is fetchedState.DONE', () => {
    const durationRenderer = getRowRenderer('duration');
    const duration = 150;
    const formatDurationSpyMockReturnValue = 'formatDurationSpyMockReturnValue';
    formatDurationSpy.mockReturnValue(formatDurationSpyMockReturnValue);

    expect(durationRenderer(duration, { state: fetchedState.ERROR })).toBe(false);
    expect(formatDurationSpy).toHaveBeenCalledTimes(0);

    expect(durationRenderer(duration, { state: fetchedState.DONE })).toBe(formatDurationSpyMockReturnValue);
    expect(formatDurationSpy).toHaveBeenCalledTimes(1);
    expect(formatDurationSpy).toHaveBeenCalledWith(duration);
  });

  it('renders link', () => {
    const linkRenderer = getRowRenderer('traceID');
    const traceID = 'trace-id';
    const renderedLink = linkRenderer(traceID);
    expect(renderedLink.type).toBe(TraceTimelineLink);
    expect(renderedLink.props).toEqual({
      traceID,
    });
  });

  it('renders NEED_MORE_TRACES_MESSAGE if cohort is too small', () => {
    expect(wrapper.contains(NEED_MORE_TRACES_MESSAGE)).toBe(false);
    updateWrapper({ cohort: cohort.slice(0, 1) });
    expect(wrapper.contains(NEED_MORE_TRACES_MESSAGE)).toBe(true);
    updateWrapper({ cohort: [] });
    expect(wrapper.contains(NEED_MORE_TRACES_MESSAGE)).toBe(true);
  });
});
