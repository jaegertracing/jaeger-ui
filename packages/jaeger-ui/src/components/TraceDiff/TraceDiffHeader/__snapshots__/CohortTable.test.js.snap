// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CohortTable renders as expected 1`] = `
Array [
  <ForwardRef(InternalTable)
    dataSource={
      Array [
        Object {
          "data": Object {
            "traceName": "trace name 0",
          },
          "error": "api error",
          "id": "trace-id-0",
          "state": "FETCH_ERROR",
        },
        Object {
          "id": "trace-id-1",
        },
        Object {
          "id": "trace-id-2",
        },
      ]
    }
    key="table"
    pagination={false}
    rowKey="id"
    rowSelection={
      Object {
        "getCheckboxProps": [Function],
        "hideDefaultSelections": true,
        "onChange": [Function],
        "selectedRowKeys": Array [
          "trace-id-0",
        ],
        "type": "radio",
      }
    }
    size="middle"
  >
    <Column
      data-testid="id"
      dataIndex="id"
      key="traceID"
      render={[Function]}
      title=""
    />
    <Column
      data-testid="traceName"
      dataIndex={
        Array [
          "data",
          "traceName",
        ]
      }
      key="traceName"
      render={[Function]}
      sortOrder="descend"
      title="Service & Operation"
    />
    <Column
      data-testid="startTime"
      dataIndex={
        Array [
          "data",
          "startTime",
        ]
      }
      key="startTime"
      render={[Function]}
      title="Date"
    />
    <Column
      data-testid="duration"
      dataIndex={
        Array [
          "data",
          "duration",
        ]
      }
      key="duration"
      render={[Function]}
      title="Duration"
    />
    <Column
      dataIndex={
        Array [
          "data",
          "spans",
          "length",
        ]
      }
      key="spans"
      title="Spans"
    />
    <Column
      className="ub-tx-center"
      data-testid="traceID"
      dataIndex={
        Array [
          "data",
          "traceID",
        ]
      }
      key="link"
      render={[Function]}
    />
  </ForwardRef(InternalTable)>,
  "",
]
`;
